README.md
# emperor-ranker
A website for people to learn about Roman Emperors and Rank them



CHANGELOG.md
# Changelog for `emperor-rank`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


test/JSONSpec.hs
module JSONSpec where

import Data.List (delete)
import JSON
import Test.Hspec (Spec, SpecWith, describe, it, shouldBe)

spec :: Spec
spec = do
  describe "watermelon" $ do
    let unObject (JObject lst) = Just lst
        unObject _ = Nothing
        theObject = unObject watermelon
        containsCucu lst = JString "Cucurbitaceae" `elem` lst
    it "length is correct" $ do
      length <$> theObject `shouldBe` Just 6
    it "first entry is genus: Citrullus" $ do
      head <$> theObject
        `shouldBe` Just ("genus", JString "Citrullus")
    it "Cucurbitaceae is in there" $ do
      fmap (containsCucu . map snd) theObject
        `shouldBe` Just True

  describe "serialize" $ do
    it "serializes null" $ do
      serialize JNull `shouldBe` "null"

    it "serializes ints" $ do
      serialize (JInt 12) `shouldBe` "12"

    it "serializes bools" $ do
      serialize (JBoolean True) `shouldBe` "true"

    it "serializes floats" $ do
      serialize (JFloat 4.58329) `shouldBe` "4.58329"

    it "serializes strings" $ do
      serialize (JString "happy happy joy joy")
        `shouldBe` "\"happy happy joy joy\""

    it "serializes arrays of ints" $ do
      serialize (JArray [JInt 1, JInt 2, JInt 3])
        `shouldBe` "[1,2,3]"

    it "serializes arrays of bools" $ do
      serialize (JArray [JBoolean False, JBoolean True])
        `shouldBe` "[false,true]"

    it "serializes objects" $ do
      serialize (JObject [("key", JString "value")])
        `shouldBe` "{\"key\":\"value\"}"

    it "serializes nested structures" $ do
      let example =
            JObject
              [ ( "dudebro",
                  JArray
                    [ JInt 4,
                      JFloat 8.75,
                      JNull
                    ]
                )
              ]
      serialize example `shouldBe` "{\"dudebro\":[4,8.75,null]}"

    it "empty structures" $ do
      serialize (JObject [("", JArray [])]) `shouldBe` "{\"\":[]}"

  describe "get" $ do
    it "gets a top-level object field" $ do
      get [O "family"] watermelon
        `shouldBe` Just (JString "Cucurbitaceae")

    it "gets a nested object field" $ do
      get [O "nutritions", O "sugar"] watermelon
        `shouldBe` Just (JInt 6)

    it "gets a field nested in objects and arrays" $ do
      let example = JObject [("smoky", JArray [JObject [("the", JObject [("bear", JBoolean False)])]])]

      get [O "smoky", A 0, O "the", O "bear"] example
        `shouldBe` Just (JBoolean False)

    it "doesn't find a field with an index that is too large" $ do
      get [O "nutritions", A 56] watermelon `shouldBe` Nothing

    it "gets a field in an array" $ do
      get [A 1] (JArray [JInt 1, JInt 2])
        `shouldBe` Just (JInt 2)

    it "doesn't find a field with an index on an object" $ do
      get [A 0] watermelon `shouldBe` Nothing

    it "doesn't allow negative indicies" $ do
      get [A (-3)] (JArray [JInt 4]) `shouldBe` Nothing

    it "doesn't allow negative indicies" $ do
      get [A 4] (JArray [JInt 4]) `shouldBe` Nothing

    it "doesn't allow indices that are too large" $ do
      let insideList = [JInt 4, JInt 5, JInt 6]
      let tooBig = fromIntegral . length $ insideList

      get [A tooBig] (JArray insideList) `shouldBe` Nothing
      get [A (tooBig - 1)] (JArray insideList)
        `shouldBe` Just (JInt 6)

  describe "jsonToFruit" $ do
    it "parses a Fruit into a JSON blob" $ do
      jsonToFruit watermelon
        `shouldBe` Just (Fruit "Watermelon" 30 0.2)
    it "refuses to parse JSON in an unexpected structure" $ do
      jsonToFruit (JInt 19) `shouldBe` Nothing

    it "parses JSON with a changed structure" $ do
      let messWithJSON (JObject lst) =
            JObject $
              ("name", JString "Watermelon By Another Name") :
              delete ("name", JString "Watermelon") lst

      jsonToFruit (messWithJSON watermelon)
        `shouldBe` Just (Fruit "Watermelon By Another Name" 30 0.2)

  describe "prettySerialize" $ do
    it "pretty serializes watermelon" $ do
      prettySerialize watermelon
        `shouldBe` "{\n    \"genus\": \"Citrullus\",\n    \"name\":\
                   \ \"Watermelon\",\n    \"id\": 25,\n    \"family\":\
                   \ \"Cucurbitaceae\",\n    \"order\": \"Cucurbitales\"\
                   \,\n    \"nutritions\": {\n        \"carbohydrates\":\
                   \ 8,\n        \"protein\": 0.6,\n        \"fat\": 0.2,\
                   \\n        \"calories\": 30,\n        \"sugar\": 6\n\
                   \    }\n}"

    it "pretty serializes nested arrays in objects" $ do
      let wut =
            JArray
              [ JObject
                  [ ( "wut",
                      JArray
                        [ JInt 1,
                          JInt 2,
                          JInt 3
                        ]
                    )
                  ],
                JInt 4
              ]
      prettySerialize wut
        `shouldBe` "[\n    {\n        \"wut\": [\n            1,\n\
                   \            2,\n            3\n        ]\n    },\
                   \\n    4\n]"

    it "pretty serializes empty arrays" $ do
      prettySerialize (JArray [])
        `shouldBe` "[]"

    it "pretty serializes arrays in objects" $ do
      prettySerialize (JObject [("", JArray [])])
        `shouldBe` "{\n    \"\": []\n}"

    it "empty objects" $ do
      prettySerialize (JObject [])
        `shouldBe` "{\n}"


src/EmperorRaw.hs
module EmperorRaw where

import Year
import Data.List

import Data.Aeson
import Data.Aeson.Types (withObject)
import Data.Text (pack)
import Control.Applicative
import Control.Monad



data EmperorRaw 
    = EmperorRaw {
                birth_cty :: String,
                name :: String,
                index:: Integer,
                name_full:: String
    }


-- instance FromJSON EmperorRaw where
--  parseJSON (Object v) =
--     EmperorRaw <$> v .: "birth_cty"
--            <*> v .: "name"
--            <*> v .: "index"
--            <*> v .: "name_full"
--  parseJSON _ = mzero

-- instance ToJSON EmperorRaw where
--  toJSON (EmperorRaw firstName lastName age likesPizza) =
--     object [ "birth_cty"  .= birth_cty
--            , "name"   .= name
--            , "index"        .= index
--            , "name_full" .= name_full
--              ]

augustus :: String
augustus = "{\"name\": \"augustus\", \"reignStart\" : 27, \"reignEnd\" : 14, \"bio\" : \"The first Roman Emperor\", \"killedBy\" : \"natural causes\"}"

augustus1 :: String
augustus1 = "{\"name\": \"augustus\", \"reignStart\" : 27, \"reignEnd\" : 14, \"bio\" : \"First Roman\"}"

augustus2 :: String
augustus2 = "{\"name\": \"augustus\", \"reignStart\" : 27}"

augustus3 :: String
augustus3 = "{\"name\": \"augustus\"}"

src/ParserJSON/JParser.hs
module ParserJSON.JParser where
import Data.Char

import ParserJSON.JSON 

import Text.Parsec hiding
  ( parse
  , choice
  , (<|>)
  , char
  , satisfy
  , eof
  , parserFail
  , sepBy
  , sepBy1
  )
import qualified Text.Parsec as P
import Text.Parsec.String (Parser)

watermelonEsc :: String
watermelonEsc = "{\"genus\":\"Citrullus\",\"name\":\"Watermelon\",\"id\":25,\"family\":\"Cucurbitaceae\",\"order\":\"Cucurbitales\",\"nutritions\":{\"carbohydrates\":[1,2,3,4],\"protein\":0.6,\"fat\":0.2,\"calories\":30,\"sugar\":6}}"

--simpler parse implementation
parse :: Parser a -> String -> Either ParseError a 
parse prsr = P.parse prsr ""

char :: Char -> Parser Char
char = P.char

satisfy :: (Char -> Bool) -> Parser Char
satisfy = P.satisfy

eof :: Parser ()
eof = P.eof

parserFail :: String -> Parser a
parserFail = P.parserFail

(<|>) :: Parser a -> Parser a -> Parser a
prsr1 <|> prsr2 = (P.<|>) (P.try prsr1) prsr2

choice :: [Parser a] -> Parser a
choice = P.choice . fmap P.try

sepBy :: Parser a -> Parser b -> Parser [a]
sepBy body sep = P.sepBy1 body (P.try sep)

--simpler sepBy1 implementation
sepBy1 :: Parser a -> Parser b -> Parser [a]
sepBy1 body sep = P.sepBy1 body (P.try sep)

augName :: String
augName = "\"name\":\"Augustus\"" 

jStartParser :: Parser Char
jStartParser = char '{'

--able to parse a JNull
jNullParser :: Parser JSON
jNullParser = (pure JNull <* string "Null")
    <|> (pure JNull <* string "null")
    <|> (pure JNull <* string "NULL")

-- able to parse a JBool
jBoolParser :: Parser JSON
jBoolParser = (pure (JBoolean True) <* string "\"true\"")
    <|> (pure (JBoolean False) <* string "\"false\"")

jStringEscapedParser :: Parser JSON
jStringEscapedParser = do
    _ <- (P.string "\"")
    input <- (P.many $ P.noneOf ['\"']) 
    _ <- (P.string "\"")
    pure $ JString input

jFullStringParser :: Parser JSON
jFullStringParser = jStringEscapedParser

jFloatParser :: Parser JSON
jFloatParser = do
   high_digits <- P.manyTill P.digit (P.char '.')
   low_digits <- P.many P.digit
   let highs = read high_digits :: Double
   let low = (read low_digits :: Double) / (10^(length low_digits))
   pure $ JFloat $ highs + low

jIntParser :: Parser JSON
jIntParser = do
   digString <- P.many P.digit
   let digits = read digString :: Int
   pure $ JInt $ digits 

   

-- building up the combined parser
jPrimitiveParse :: Parser JSON
jPrimitiveParse =  jNullParser 
    <|> jBoolParser
    <|> jObjectParser
    <|> jArryParser
    <|> jFullStringParser
    <|> jFloatParser
    <|> jIntParser


jParse :: Parser JSON
jParse = jPrimitiveParse
  

--parser to parse a command conjuction
jsonParseCombined :: Parser [JSON]
jsonParseCombined = do 
    jsons <- (sepBy1 (jParse) (P.string ","))
    _ <- P.eof
    pure jsons

parseString :: Parser String
parseString = do
    _ <- (P.string "\"")
    input <- (P.many $ P.noneOf ['\"']) 
    _ <- (P.string "\"")
    pure input


keyValueSeperator :: P.Parsec String () ()
keyValueSeperator = do
    P.spaces
    P.char ':'
    P.spaces

jPairParser :: Parser (String, JSON)
jPairParser = do
  _ <- P.spaces
  nm <- parseString
  _ <- keyValueSeperator
  dt <- jPrimitiveParse
  _ <- P.spaces
  pure (nm, dt)
 
pairSeparator :: P.Parsec String () ()
pairSeparator = do
    P.spaces
    P.char ','
    P.spaces

jObjectParser :: Parser JSON
jObjectParser = do
   _ <- char '{'
   parsedObject <-  sepBy jPairParser (pairSeparator) 
   _ <- char '}'
   pure $ JObject parsedObject

jArryParser :: Parser JSON
jArryParser = do 
   _ <- char '['
   parsedObject <-  sepBy jPrimitiveParse (pairSeparator) 
   _ <- char ']'
   pure $ JArray parsedObject

parseInput :: String -> Maybe JSON
parseInput str = case (parsed) of
                  Right x -> Just x
                  Left _ -> Nothing
    where parsed = (do
            jsons <- parse jPrimitiveParse str
            pure (jsons))
    



readJsonFile :: String -> IO String
readJsonFile nm = do
   content <- readFile $ "./data/" ++  nm ++ ".json"
   pure content 

-- parseJsonFromFile :: String -> Maybe JSON 
-- parseJsonFromFile nm = do
--         text <- readJsonFile nm
--         return $ JNull




src/ParserJSON/JSON.hs
module ParserJSON.JSON 
where

import Data.List 

import Text.Parsec hiding (parse, runParser, (<|>), sepBy1, choice)
import qualified Text.Parsec as P 
import Text.Parsec.String (Parser)
 
-- Exercise 1: watermelon
data JSON
  = JNull
  | JBoolean Bool
  | JInt Int
  | JFloat Double
  | JString String
  | JArray [JSON]
  | JObject [(String, JSON)]
  deriving (Show, Eq)


watermelon :: JSON
watermelon = JObject [
	   ("genus", JString "Citrullus"),
	   ("name", JString "Watermelon"),
	   ("id", JInt 25),
	   ("family", JString "Cucurbitaceae"),
	   ("order", JString "Cucurbitales"),
	   ("nutritions", JObject [
	     ("carbohydrates", JInt 8),
	     ("protein", JFloat 0.6),
	     ("fat", JFloat 0.2),
	     ("calories", JInt 30),
	     ("sugar", JInt 6)
	   ])
	 ]


nutrientList :: JSON
nutrientList = JObject [
	   ("nutritions", JObject [
	     ("carbohydrates", JInt 8),
	     ("protein", JFloat 0.6),
	     ("fat", JFloat 0.2),
	     ("calories", JInt 30),
	     ("sugar", JInt 6)
	   ])
	 ]

carbs :: (String, JSON)
carbs = ("carbohydrates", JArray [JInt 1, JInt 2, JInt 3])

  -- = JNull
  -- | JBoolean Bool
  -- | JInt Int
  -- | JFloat Double
  -- | JString String
  -- | JArray [JSON]
  -- | JObject [(String, JSON)]

quote :: String -> String
quote str = "\"" ++ str ++ "\"" 

-- Exercise 2: serialize
serialize :: JSON -> String
serialize (JString x) = show x
serialize (JBoolean x) = if x then "true" else "false"
serialize (JInt x) = show x
serialize (JFloat x) = show x
serialize (JNull) = "null"
serialize (JObject lst) = serialListHelper lst
serialize (JArray arr) = "[" ++ (commaify $ map serialize arr) ++ "]"

--take list of string and turn into a comma seperated string
commaify :: [String] -> String
commaify lst =
      foldl (\acc item -> if acc == "" then (item) else (item) ++ "," ++ acc) "" (reverse lst)

serialTupleHelper :: (String, JSON) -> String
serialTupleHelper (key, val) = "{" ++ (quote key) ++ ":" ++ (serialize val) ++ "}"

serialListHelper :: [(String, JSON)] -> String
serialListHelper lst = commaify listed
        where listed =  map (\x -> serialTupleHelper x) lst
      


-- Exercise 3: get
data NextStep
  = O String
  | A Int
  deriving Show

--   λ> get [O "family"] watermelon
-- Just (JString "Cucurbitaceae")
-- λ> get [O "nutritions", O "sugar"] watermelon
-- Just (JInt 6)


--I know the syntax is bad, but I'm truong t 
get :: [NextStep] -> JSON -> Maybe JSON
get = undefined  --_ -> JObject lst -> map (isMatch) lst


isMatch :: String -> (String, JSON) -> Maybe JSON
isMatch name (key, value) 
    | (name == key ) = Just value
    | otherwise = Nothing

--trying to find the match in the JSon object
findOnObject ::  String -> JSON -> Maybe JSON
findOnObject str (JObject lst) = case filtered of
         [] -> Nothing
         otherwise -> head filtered
      where checkList =  [(isMatch str)] <*> lst
            filtered = filter (\x -> x /= Nothing) checkList

--just straight didn't get to this
-- Exercise 4: jsonToFruit
data Fruit = Fruit { name :: String
                   , calories :: Int
                   , fat :: Double
                   }
             deriving (Show, Eq)

jsonToFruit :: JSON -> Maybe Fruit
jsonToFruit = undefined


-- Extra Credit: prettySerialize
prettySerialize :: JSON -> String
prettySerialize = undefined

deserializePure :: String -> JSON
deserializePure = undefined

watermelonString :: String
watermelonString = "{\"genus\":\"Citrullus\"},{\"name\":\"Watermelon\"},{\"id\":25},{\"family\":\"Cucurbitaceae\"},{\"order\":\"Cucurbitales\"},{\"nutritions\":{\"carbohydrates\":8},{\"protein\":0.6},{\"fat\":0.2},{\"calories\":30},{\"sugar\":6}}"


  
--test
jstringTuple :: (String, String) -> (String, JSON)
jstringTuple inp =  fmap (\x -> JString x) inp

jdoubleTuple :: (String, Double) -> (String, JSON)
jdoubleTuple inp =  fmap (\x -> JFloat x) inp



stringList1 = [("genus","Citrullus"),("name", "Watermelon")]
doubleList = [("id", 25)]
stringList2 = [("family", "Cucurbitaceae"),("order", "Cucurbitales")]
testNutrientList = [("carbohydrates", 8), ("protein",0.6), ("fat", 0.2), ("calories",30),  ("sugar",6)]

makeNutrient :: (String, Double) -> (String, JSON)
makeNutrient (name,value) = (name, JFloat value)


makeNutrientList :: [(String, Double)] -> JSON
makeNutrientList lst =  JObject rawList
      where rawList = map (\item -> makeNutrient item) lst


testList10 = fmap (\x -> jstringTuple x) stringList1